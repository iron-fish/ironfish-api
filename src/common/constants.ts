/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { EventType } from '.prisma/client';

export const DAYS_IN_WEEK = 7;
export const MS_PER_DAY = 1000 * 60 * 60 * 24;

// Pagination limits
export const DEFAULT_LIMIT = 20;
export const MAX_LIMIT = 100;

// Event categories
export const WEEKLY_POINT_LIMITS_BY_EVENT_TYPE: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 1000,
  [EventType.BUG_CAUGHT]: 1000,
  [EventType.COMMUNITY_CONTRIBUTION]: 2000,
  [EventType.PULL_REQUEST_MERGED]: 5000,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 1000,
};

export const POINTS_PER_CATEGORY: Record<EventType, number> = {
  [EventType.BLOCK_MINED]: 100,
  [EventType.BUG_CAUGHT]: 100,
  [EventType.COMMUNITY_CONTRIBUTION]: 1000,
  [EventType.PULL_REQUEST_MERGED]: 500,
  [EventType.SOCIAL_MEDIA_PROMOTION]: 100,
};
