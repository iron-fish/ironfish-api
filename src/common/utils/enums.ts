/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

export type StrEnumValue<T> = T[keyof T];
export type StrEnum<T> = Record<keyof T, string>;

function getEnumStrValues<T extends StrEnum<T>>(
  enumType: T,
): Array<StrEnumValue<T>> {
  return Object.values(enumType)
    .filter((v) => typeof v === 'string')
    .map((v) => v as StrEnumValue<T>);
}

export function isInStrEnum<T extends StrEnum<T>>(
  value: unknown,
  enumType: T,
): boolean {
  return !!getEnumStrValues(enumType).find((enumValue) => enumValue === value);
}
