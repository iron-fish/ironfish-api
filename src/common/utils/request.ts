/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Request } from 'express';

export function fetchIpAddressFromRequest(request: Request): string {
  const xForwardedFor = request.header('X-Forwarded-For');
  if (xForwardedFor) {
    const addresses = xForwardedFor.split(',');
    if (addresses.length) {
      return addresses[0].trim();
    }
  }

  // In Heroku, `X-Forwarded-For` will always be set. However, this will be used
  // as a fallback in local development.
  return request.ip;
}
