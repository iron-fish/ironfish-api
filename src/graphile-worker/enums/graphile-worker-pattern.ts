/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
export enum GraphileWorkerPattern {
  DELETE_BLOCK_MINED_EVENT = 'DELETE_BLOCK_MINED_EVENT',
  CREATE_NODE_UPTIME_EVENT = 'CREATE_NODE_UPTIME_EVENT',
  REFRESH_USERS_POINTS = 'REFRESH_USERS_POINTS',
  REFRESH_USER_POINTS = 'REFRESH_USER_POINTS',
  SYNC_BLOCKS_DAILY = 'SYNC_BLOCKS_DAILY',
  UPDATE_LATEST_POINTS = 'UPDATE_LATEST_POINTS',
  UPSERT_BLOCK_MINED_EVENT = 'UPSERT_BLOCK_MINED_EVENT',
}
