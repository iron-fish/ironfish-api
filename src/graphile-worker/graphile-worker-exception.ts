/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Catch, RpcExceptionFilter } from '@nestjs/common';
import { Observable } from 'rxjs';

@Catch()
export class GraphileWorkerException implements RpcExceptionFilter<Error> {
  catch(error: Error): Observable<unknown> {
    throw error;
  }
}
