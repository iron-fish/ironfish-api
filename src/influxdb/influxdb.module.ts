/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Module } from '@nestjs/common';
import { ApiConfigModule } from '../api-config/api-config.module';
import { InfluxDbService } from './influxdb.service';

@Module({
  exports: [InfluxDbService],
  imports: [ApiConfigModule],
  providers: [InfluxDbService],
})
export class InfluxDbModule {}
