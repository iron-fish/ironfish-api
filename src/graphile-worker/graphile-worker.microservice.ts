/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { CustomTransportStrategy, Server } from '@nestjs/microservices';
import { run, Runner, Task, TaskList } from 'graphile-worker';
import { Pool, PoolConfig } from 'pg';
import { ApiConfigService } from '../api-config/api-config.service';
import { DatadogService } from '../datadog/datadog.service';
import { LoggerService } from '../logger/logger.service';
import { GraphileWorkerPattern } from './enums/graphile-worker-pattern';
import { GraphileWorkerHandler } from './types/graphile-worker-handler';

export class GraphileWorkerMicroservice
  extends Server
  implements CustomTransportStrategy
{
  private runner!: Runner;

  constructor(
    private readonly config: ApiConfigService,
    private readonly datadogService: DatadogService,
    private readonly loggerService: LoggerService,
  ) {
    super();
  }

  async listen(): Promise<void> {
    this.runner = await run({
      noHandleSignals: false,
      pgPool: new Pool(this.getPostgresPoolConfig()),
      taskList: this.getTaskHandlers(),
    });
  }

  async close(): Promise<void> {
    await this.runner.stop();
  }

  private getPostgresPoolConfig(): PoolConfig {
    return {
      connectionString: this.config.get<string>('DATABASE_URL'),
      ssl:
        this.config.get<string>('NODE_ENV') !== 'development'
          ? { rejectUnauthorized: false }
          : undefined,
    };
  }

  private getTaskHandlers(): TaskList {
    const patterns = Array.from(this.messageHandlers.keys());
    const taskList: Record<string, Task> = {};
    for (const pattern of patterns) {
      taskList[pattern] = this.createMessageHandler(
        pattern as unknown as GraphileWorkerPattern,
      );
    }
    return taskList;
  }

  private createMessageHandler(pattern: GraphileWorkerPattern): Task {
    return async (payload: unknown) => {
      const start = new Date().getTime();
      try {
        await this.handle(pattern, payload);

        this.datadogService.increment('worker.success', 1, {
          pattern,
        });
      } catch (error) {
        if (error instanceof Error) {
          this.datadogService.increment('worker.error', 1, {
            pattern,
            type: error.constructor.name,
          });
        }

        throw error;
      } finally {
        const duration = new Date().getTime() - start;

        this.datadogService.timing('worker', duration, {
          pattern,
        });

        this.loggerService.info(
          JSON.stringify({
            duration,
            pattern,
          }),
        );
      }
    };
  }

  private async handle(
    pattern: GraphileWorkerPattern,
    payload: unknown,
  ): Promise<void> {
    const handler = this.messageHandlers.get(
      pattern.toString(),
    ) as GraphileWorkerHandler;
    if (!handler) {
      throw new Error(`Undefined pattern: '${pattern}'`);
    }

    const response = await handler(payload as Record<string, unknown>);
    if (response && response.requeue) {
      throw new Error('Retry Graphile Worker message');
    }
  }
}
