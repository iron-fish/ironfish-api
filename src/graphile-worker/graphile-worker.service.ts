/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { Injectable } from '@nestjs/common';
import { Job, makeWorkerUtils, WorkerUtils } from 'graphile-worker';
import { Pool, PoolConfig } from 'pg';
import { v4 as uuid } from 'uuid';
import { ApiConfigService } from '../api-config/api-config.service';
import { GraphileWorkerPattern } from './enums/graphile-worker-pattern';
import { GraphileWorkerJobOptions } from './interfaces/graphile-worker-job-options';

@Injectable()
export class GraphileWorkerService {
  private workerUtils!: WorkerUtils;
  private workerUtilsPromise: Promise<WorkerUtils> | null = null;

  constructor(private readonly config: ApiConfigService) {}

  async addJob<T = unknown>(
    pattern: GraphileWorkerPattern,
    payload?: T,
    { queueName, runAt, jobKey, jobKeyMode }: GraphileWorkerJobOptions = {},
  ): Promise<Job> {
    await this.initWorkerUtils();

    return this.workerUtils.addJob(pattern.toString(), payload, {
      jobKey: jobKey || `job_${uuid()}`,
      jobKeyMode,
      queueName,
      runAt,
    });
  }

  async queuedJobCount(): Promise<number> {
    await this.initWorkerUtils();

    return this.workerUtils.withPgClient(async (pgClient) => {
      const result = await pgClient.query<{ count: bigint }>(
        'SELECT COUNT(*) FROM graphile_worker.jobs WHERE locked_at IS NULL;',
      );

      return Number(result.rows[0].count);
    });
  }

  private async initWorkerUtils(): Promise<void> {
    if (this.workerUtils) {
      return;
    }

    if (this.workerUtilsPromise) {
      await this.workerUtilsPromise;
      return;
    }

    this.workerUtilsPromise = makeWorkerUtils({
      pgPool: new Pool(this.getPostgresPoolConfig()),
    });

    this.workerUtils = await this.workerUtilsPromise;
  }

  private getPostgresPoolConfig(): PoolConfig {
    return {
      connectionString: this.config.get<string>('DATABASE_URL'),
      ssl:
        this.config.get<string>('NODE_ENV') !== 'development'
          ? { rejectUnauthorized: false }
          : undefined,
    };
  }
}
