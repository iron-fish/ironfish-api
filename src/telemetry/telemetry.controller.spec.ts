/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import { HttpStatus, INestApplication } from '@nestjs/common';
import faker from 'faker';
import request from 'supertest';
import { v4 as uuid } from 'uuid';
import { GraphileWorkerPattern } from '../graphile-worker/enums/graphile-worker-pattern';
import { GraphileWorkerService } from '../graphile-worker/graphile-worker.service';
import { InfluxDbService } from '../influxdb/influxdb.service';
import { NodeUptimesService } from '../node-uptimes/node-uptimes.service';
import { PrismaService } from '../prisma/prisma.service';
import { bootstrapTestApp } from '../test/test-app';

describe('TelemetryController', () => {
  let app: INestApplication;
  let graphileWorkerService: GraphileWorkerService;
  let influxDbService: InfluxDbService;
  let nodeUptimesService: NodeUptimesService;
  let prisma: PrismaService;

  beforeAll(async () => {
    app = await bootstrapTestApp();
    graphileWorkerService = app.get(GraphileWorkerService);
    influxDbService = app.get(InfluxDbService);
    nodeUptimesService = app.get(NodeUptimesService);
    prisma = app.get(PrismaService);
    await app.init();
  });

  afterAll(async () => {
    await app.close();
  });

  describe('POST /telemetry', () => {
    describe('with missing arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post('/telemetry')
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with empty or invalid arguments', () => {
      it('returns a 422', async () => {
        const { body } = await request(app.getHttpServer())
          .post('/telemetry')
          .send({
            points: [
              {
                fields: [
                  {
                    name: 'foo',
                    type: 'missing-type',
                    value: 'howdy',
                  },
                  {
                    name: 'foo',
                    type: 'boolean',
                  },
                  {
                    name: 'foo',
                    type: 'float',
                    value: 'ironfish',
                  },
                  {
                    name: 'foo',
                    type: 'integer',
                    value: 'hello',
                  },
                  {
                    name: 'foo',
                    type: 'string',
                    value: 123,
                  },
                ],
                measurement: '',
                tags: [{ foo: 'bar' }],
                value: -1,
              },
            ],
          })
          .expect(HttpStatus.UNPROCESSABLE_ENTITY);

        expect(body).toMatchSnapshot();
      });
    });

    describe('with a version lower than the minimum telemetry version', () => {
      it('does not write the points to InfluxDB', async () => {
        const writePoints = jest
          .spyOn(influxDbService, 'writePoints')
          .mockImplementationOnce(jest.fn());
        const fields = [
          {
            name: 'online',
            type: 'boolean',
            value: true,
          },
          {
            name: 'memory',
            type: 'float',
            value: 1.23,
          },
          {
            name: 'mempool',
            type: 'integer',
            value: 1,
          },
          {
            name: 'name',
            type: 'string',
            value: 'howdy',
          },
        ];
        const measurement = 'node';
        const tags = [{ name: 'version', value: '0.0.0' }];

        await request(app.getHttpServer())
          .post('/telemetry')
          .send({
            points: [{ fields, measurement, tags, timestamp: new Date() }],
          })
          .expect(HttpStatus.CREATED);

        expect(writePoints).toHaveBeenCalledTimes(0);
      });
    });

    describe('with valid arguments', () => {
      it('writes the point to InfluxDB', async () => {
        const writePoints = jest
          .spyOn(influxDbService, 'writePoints')
          .mockImplementationOnce(jest.fn());
        const fields = [
          {
            name: 'online',
            type: 'boolean',
            value: true,
          },
          {
            name: 'memory',
            type: 'float',
            value: 1.23,
          },
          {
            name: 'mempool',
            type: 'integer',
            value: 1,
          },
          {
            name: 'name',
            type: 'string',
            value: 'howdy',
          },
        ];
        const measurement = 'node';
        const tags = [{ name: 'version', value: '0.1.24' }];
        const timestamp = new Date();

        await request(app.getHttpServer())
          .post('/telemetry')
          .send({ points: [{ fields, measurement, tags, timestamp }] })
          .expect(HttpStatus.CREATED);

        expect(writePoints).toHaveBeenCalledTimes(1);
        expect(writePoints).toHaveBeenCalledWith([
          {
            fields,
            measurement,
            tags,
            timestamp,
          },
        ]);
      });

      it('updates the node uptime', async () => {
        const nodeUptimeUpsert = jest
          .spyOn(nodeUptimesService, 'addUptime')
          .mockImplementationOnce(jest.fn());

        const graffiti = uuid();
        const user = await prisma.user.create({
          data: {
            email: faker.internet.email(),
            graffiti,
            country_code: faker.address.countryCode(),
          },
        });

        await request(app.getHttpServer())
          .post('/telemetry')
          .send({ points: [], graffiti })
          .expect(HttpStatus.CREATED);

        expect(nodeUptimeUpsert).toHaveBeenCalledTimes(1);
        expect(nodeUptimeUpsert).toHaveBeenCalledWith(user);
      });

      it('updates users points when user has logged enough hours', async () => {
        const workerAddJob = jest
          .spyOn(graphileWorkerService, 'addJob')
          .mockImplementationOnce(jest.fn());

        const oldCheckin = new Date();
        oldCheckin.setHours(oldCheckin.getHours() - 2);

        const user = await prisma.user.create({
          data: {
            email: faker.internet.email(),
            graffiti: uuid(),
            country_code: faker.address.countryCode(),
          },
        });

        await prisma.nodeUptime.create({
          data: {
            user_id: user.id,
            total_hours: 12,
            last_checked_in: oldCheckin,
          },
        });

        await request(app.getHttpServer())
          .post('/telemetry')
          .send({ points: [], graffiti: user.graffiti })
          .expect(HttpStatus.CREATED);

        expect(workerAddJob).toHaveBeenCalledTimes(1);
        expect(workerAddJob).toHaveBeenCalledWith(
          GraphileWorkerPattern.CREATE_NODE_UPTIME_EVENT,
          { userId: user.id },
        );
      });
    });
  });
});
