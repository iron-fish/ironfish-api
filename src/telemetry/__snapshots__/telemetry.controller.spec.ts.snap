// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TelemetryController POST /telemetry with empty or invalid arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "points.0.fields.0.type must be one of the following values: boolean, float, integer, string",
    "points.0.fields.1.value must be a boolean value",
    "points.0.fields.2.value must be a positive number",
    "points.0.fields.2.value must be a number conforming to the specified constraints",
    "points.0.fields.3.value must be a positive number",
    "points.0.fields.3.value must be an integer number",
    "points.0.fields.4.value must be a string",
    "points.0.measurement should not be empty",
    "points.0.timestamp must be a Date instance",
    "points.0.tags.0.name must be a string",
    "points.0.tags.0.name should not be empty",
    "points.0.tags.0.value must be a string",
    "points.0.tags.0.value should not be empty",
  ],
  "statusCode": 422,
}
`;

exports[`TelemetryController POST /telemetry with missing arguments returns a 422 1`] = `
Object {
  "error": "Unprocessable Entity",
  "message": Array [
    "points must contain not more than 3000 elements",
    "points must be an array",
  ],
  "statusCode": 422,
}
`;
